aopy/precondition/base.py

```diff
@@ -393,6 +395,95 @@ def detect_spikes(spike_filt_data, samplerate, threshold, above_thresh=True,  wf
 
     return spike_times, spike_waveforms
 
+def detect_spikes_chunk(spike_filt_data, samplerate, threshold, chunksize, above_thresh=True,  wf_length=1000, tbefore_wf=200):
+    '''
+    This function is based on detect_spikes and calculates spike times while dividing data into chunks to deal with memory issues.
+    This may work for numpy memmap array
+
+    Args:
+        spike_filt_data (nt,nch): Time series neural data to detect spikes and extract waveforms from.
+        samplerate (float): Sampling rate [Hz]
+        threshold (float): Threshold that input data must cross to indicate a spike for each channel.
+        chunksize (int): Chunk size to process data by chunk
+        above_thresh (bool): If True, only spikes above the threshold will be detected. If false, only spikes below threshold will be detected. 
+        wf_length (float): Length of waveforms to output :math:`[\mu s]`. Actual length will be rounded up. If set to 'None', waveforms will not be returned. Defaults to 1000 :math:`\mu s`
+        tbefore_wf (float): Length of waveform to include before spike detection time :math:`[\mu s]`. Defaults to 200 :math:`\mu s`  
+        
+    Returns: 
+        tuple: Tuple containing:
+            | **spike_times (list of spike times):**  List of nspike length arrays with each list element corresponding to a channel. Spike times are in seconds.
+            | **spike_waveforms (list of waveforms):** List of (nspike, nwf_pts) arrays with each list element corresponding to a channel. Returns NaN if there aren't enough data points to retreive a full waveform.
+    '''
+    if spike_filt_data.ndim == 1:
+        spike_filt_data = spike_filt_data.reshape(-1,1)
+        
+    nt,nch = spike_filt_data.shape
+    nchunk = math.ceil(nt/chunksize)
+    chunktime = chunksize/samplerate
+    
+    # Calculate spike times
+    n_samples = 0
+    previous_thresh = np.ones(nch) == 1
+    spike_times = []
+    spike_waveforms =[]
+    
+    for ichunk in range(nchunk):
+        if ichunk == nchunk:
+            data = spike_filt_data[n_samples:,:]
+        else:
+            data = spike_filt_data[n_samples:n_samples+chunksize,:]
+
+        if above_thresh:
+            data_above_thresh_mask = data > threshold
+        else:        
+            data_above_thresh_mask = data < threshold
+
```

Comment:
We probably want this to output data either above or below the threshold. It probably isn't necessary to make this change in this pull request but we should open an issue detailing it.

Reply:
Actually, can we just use the absolute value of the signal and use a positive threshold?